src/FindEmailAddress.java

```diff
@@ -0,0 +1,362 @@
+    public String requestDomain;
+    public Hashtable<String,String> urlHistory;
+    public Hashtable<String,String> emails;
+    public static int maxUrlCheck = 200; //to avoid too many pages to check we add maximum number of url to check, it can be modified
+    public ExecutorService executor;
+    public Integer tasksNumber;
+
+    public static void main(String[] args)
+    {
+        String requestUrl = null;
+
+        if(args.length >0)
+        {
+            requestUrl =  args[0];
+        }
+        if (requestUrl == null)
+            return;
+
+        FindEmailAddress findEmailAddress = new FindEmailAddress(requestUrl);
+        findEmailAddress.startWebCrawl();
+        findEmailAddress.printout();
+    }
+
+    /*
+    *  @param requestURL is the root url. All the webpage with same domain will be searched for email address
+    */
+    public FindEmailAddress(String requestURL)
+    {
+        this.rootURL = requestURL;
+        if(!requestURL.toLowerCase().startsWith("http"))
+        {
+            rootURL = "http://" + requestURL;
+        }
+        try {
+            URI uri = new URI(rootURL);
+            this.requestDomain= uri.getHost();
+
+        }catch(java.net.URISyntaxException e)
+        {
+            e.printStackTrace();
+        }
+        this.emails = new Hashtable<String,String>();
+        this.urlHistory = new Hashtable<String, String>();
+        this.urlHistory.put(rootURL,"");
+        tasksNumber = 0;
+    }
+
+    /*
+     *  This is the function to do web crawling
+     *  it will create a thread pool
+     *  if all the tasks are done, it will stop.
+     */
+
+    public void startWebCrawl()
+    {
+        if(rootURL == null || rootURL.trim().length()<=0)
+            return;
+
+        executor = Executors.newFixedThreadPool(10);
+        executor.execute(new Handler(rootURL));
+        do{
+            try {
+                Thread.sleep(3000);
+            }catch(Exception e)
+            {
+                e.printStackTrace();
+            }
+        }while(this.tasksNumber >0);
+
+        executor.shutdown();
+    }
+
+/*
+* this inline class is for http request and email check running task
+*/
+    class Handler implements Runnable {
+        String requestURL;
+        Handler(String requestURL) {
+            this.requestURL = requestURL;
+        }
+        public void run() {
+            System.out.println("start search url :" + requestURL);
+            makeHttpReuqest(requestURL);
+        }
+    }
+
+    //need multithread
+    /*
+    * This is the function to make http call.
+    * And this function will also do the email search and same domain url search.
+    * the email address will be saved in a hashtable
+    * the urls will be saved in a vector, and a history tashtable is for filtering the urls already visited.
+    */
+    private void makeHttpReuqest(String httpURLString)
+    {
+        if(httpURLString == null || httpURLString.trim().length()<=0)
+            return;
```

Comment:
this can also cause problems, you must decrement `tasksNumber` if you're going to return early here.

Reply:
Yes. You are right, need to be fixed.  Although in this project it will not happen (the httpURLString will not be null or "").